src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -3039,6 +3039,14 @@ public override BoundNode VisitConditionalReceiver(BoundConditionalReceiver node
             return null;
         }
 
+        public override BoundNode VisitLoweredIsPatternExpression(BoundLoweredIsPatternExpression node)
+        {
+            var savedState = this.State.Clone();
+            VisitStatements(node.Statements);
+            SetState(savedState);
```

Comment:
>SetState(savedState); [](http://example.com/codeflow?start=12&length=21)

This doesn't look right. I think that here we also should act as though we have two implicit label statements and an implicit goto to the end. I am thinking about code doing something like:
```
VisitLabel(node.WhenTrueLabel);
var stateWhenTrue = this.State.Clone();
SetUnreachable();
VisitLabel(node.WhenFalseLabel);
Join(ref this.State, ref stateWhenTrue);
```

Reply:
>we have two implicit label statements and an implicit goto to the end

That doesn't quite reflect the final flow,

https://github.com/dotnet/roslyn/blob/283061a27a5994d90190bb98e501f992399da4c0/src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs#L370-L377

Also there's no such overload for `VisitLabel(node.WhenTrueLabel);` we do have `VisitLabel(LabelSymbol, BoundStatement)` but we're working with an expression and I don't think you meant to change that method?

Reply:
Sorry I misread "to the end" as "at the end", but still I'm not sure what exactly should VisitLabel do.

Reply:
Another question, since we're working with constants and known labels, what `SetUnreachable` signifies here? Handling of the labels should supposedly cover the unreachable state, in case, for example, there were no jumps to either labels.

Reply:
Final note, I think this method should be an inlined version of the same impl for switch in the other PR.
```cs
return e is p;
// same as the following for this analysis
return e switch { p => true, _ => false };
```
Not saying that either of these are correct at this time, but they should definitely match. I believe this can help to verify the implementation in either PR.

https://github.com/dotnet/roslyn/blob/755afc6c7fb4217da5d2dc1b4f712280151bc3c1/src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass_Switch.cs#L185-L200

First off, notice we append binding statements to `LoweredIs.Statements`, but for switch the state is reset beforehand. 

For reference, at the moment `LoweredSwitchArm.Statements` contains:

1. lowered where clause including the exit jump if it evaluates to false, if any
2. binding statements, if any (this is the collection returned from BuilderForSection for both rewriters)
3. switch arm label statement
4. spilled value produced in spiller, if any

And finally the the value is emitted which is just a constant for is-expressions.

Reply:
> Also there's no such overload for `VisitLabel(node.WhenTrueLabel);` we do have `VisitLabel(LabelSymbol, BoundStatement)` but we're working with an expression and I don't think you meant to change that method?

This is how that local function should look like in my opinion:
```
        void visitLabel(LabelSymbol label)
        {
            ResolveBranches(label, null);
            var state = LabelState(label);
            Join(ref this.State, ref state);
            _labels[label] = this.State.Clone();
        }
```

> Another question, since we're working with constants and known labels, what `SetUnreachable`/`Join` signify here? 

`SetUnreachable` signifies that ```node.WhenFalseLabel``` is not reachable in case of a jump to ```node.WhenTrueLabel```.

`Join` signifies that control flow merges at the end, see what we do in ```EmitLoweredIsPatternExpression```. 

> Final note, I think this method should be an inlined version of the same impl for switch in the other PR which can be seen as a generalized `is` expression:

Let's focus on the simplified version of the problem that we have at hand, and get it right first. We will deal with problems in other PRs when we get to other PRs.

Reply:
>This is how that local function should look

Thanks, didn't know the statement parameter is completely optional.